File: .reasonix/skills/abacatepay/examples/go/webhook.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"time"
)

var processedEvents = make(map[string]bool) // Simple in-memory for idempotency

func webhookHandler(w http.ResponseWriter, r *http.Request) {
	secret := os.Getenv("ABACATEPAY_WEBHOOK_SECRET")

	querySecret := r.URL.Query().Get("webhookSecret")
	if querySecret != secret {
		fmt.Println("Invalid secret query")
		http.Error(w, "Invalid secret query", http.StatusUnauthorized)
		return
	}

	bodyBytes, err := io.ReadAll(r.Body)
	if err != nil {
		fmt.Println("Error reading body:", err)
		http.Error(w, "Error reading body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	// Basic validation
	var event map[string]interface{}
	if err := json.Unmarshal(bodyBytes, &event); err != nil || event["id"] == nil {
		fmt.Println("Invalid payload structure")
		http.Error(w, "Invalid payload", http.StatusBadRequest)
		return
	}

	timestampStr := r.Header.Get("X-Webhook-Timestamp")
	if timestampStr == "" {
		fmt.Println("Missing timestamp")
		http.Error(w, "Missing timestamp", http.StatusUnauthorized)
		return
	}
	timestamp, err := strconv.ParseInt(timestampStr, 10, 64)
	if err != nil || time.Now().Unix()-timestamp > 300 { // 5 minutes
		fmt.Println("Invalid or expired timestamp")
		http.Error(w, "Invalid timestamp", http.StatusUnauthorized)
		return
	}

	signature := r.Header.Get("X-Webhook-Signature")

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(bodyBytes)
	expectedSignature := base64.StdEncoding.EncodeToString(mac.Sum(nil))

	// Constant-time comparison
	if subtle.ConstantTimeCompare([]byte(signature), []byte(expectedSignature)) != 1 {
		fmt.Println("Invalid signature")
		http.Error(w, "Invalid signature", http.StatusUnauthorized)
		return
	}

	eventID := event["id"].(string)
	eventType := event["event"].(string)

	// Idempotency
	if processedEvents[eventID] {
		fmt.Println("Event already processed:", eventID)
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("Already processed"))
		return
	}

	fmt.Println("Processing event:", eventType, "ID:", eventID)

	// Simulate processing
	if eventType == "billing.paid" {
		// Process payment
		fmt.Println("Payment confirmed for:", event["data"].(map[string]interface{})["id"])
		processedEvents[eventID] = true
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("OK"))
	} else {
		// Unknown event, but accept to avoid retries
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("OK"))
	}
}

File: .reasonix/skills/abacatepay/examples/ts/webhook.ts
import express from 'express';
import crypto from 'crypto';

const app = express();

app.use(express.json({
    verify: (req: any, res, buf) => {
        req.rawBody = buf;
    }
}));

const WEBHOOK_SECRET = process.env.ABACATEPAY_WEBHOOK_SECRET;

app.post('/webhooks/abacatepay', async (req, res) => {
    const signature = req.headers['x-webhook-signature'] as string;
    const timestamp = req.headers['x-webhook-timestamp'] as string;
    const secretFromQuery = req.query.webhookSecret as string;

    // Validation
    if (!req.body || !req.body.id) {
        console.error('Invalid payload structure');
        return res.status(400).send('Invalid payload');
    }

    if (secretFromQuery !== WEBHOOK_SECRET) {
        console.error('Invalid secret query');
        return res.status(401).send('Invalid secret query');
    }

    // Timestamp check to prevent replay attacks
    if (!timestamp || Math.abs(Date.now() / 1000 - parseInt(timestamp)) > 300) { // 5 minutes
        console.error('Invalid or expired timestamp');
        return res.status(401).send('Invalid timestamp');
    }

    const hmac = crypto.createHmac('sha256', WEBHOOK_SECRET as string);
    hmac.update(req.rawBody);
    const expectedSignature = hmac.digest('base64');

    // Constant-time comparison to prevent timing attacks
    if (!crypto.timingSafeEqual(Buffer.from(signature, 'base64'), Buffer.from(expectedSignature, 'base64'))) {
        console.error('Invalid signature');
        return res.status(401).send('Invalid signature');
    }

    const { event: eventType, id, data } = req.body;
    console.log('Received event:', eventType, 'ID:', id);

    try {
        // Idempotency check (assume a processedEvents set or DB)
        if (processedEvents.has(id)) {
            console.log('Event already processed:', id);
            return res.status(200).send('Already processed');
        }

        if (eventType === 'billing.paid') {
            console.log('Payment confirmed for:', data.id);
            // Process payment logic
            processedEvents.add(id);
        }

        res.status(200).send('OK');
    } catch (error: any) {
        console.error('Error processing webhook:', error.message);
        // For failures, return 500 to trigger retry
        res.status(500).send('Processing failed');
    }
});